cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,540 @@
+                     j += local_size) {
+                    bool exists = j < group_aligned_size || local_id < rem;
+                    std::uint8_t c = exists ? cinput[j * radix_count + i] : 0;
+                    std::uint32_t entry = 0;
+                    bool entry_found = false;
+                    for (std::uint32_t k = 0; k < local_size; k++) {
+                        bool correct = j < group_aligned_size || k < rem;
+                        std::uint32_t done = sycl::ONEAPI::broadcast(sbg, correct ? 0 : 1, k);
+                        if (done)
+                            break;
+                        std::uint8_t value = sycl::ONEAPI::broadcast(sbg, c, k);
+                        if (!entry_found && value == c) {
+                            entry = k;
+                            entry_found = true;
+                        }
+                        IntType count =
+                            sycl::ONEAPI::reduce(sbg,
+                                                 static_cast<IntType>(exists && value == c ? 1 : 0),
+                                                 cl::sycl::ONEAPI::plus<IntType>());
+                        if (entry_found && entry == local_id && entry == k) {
+                            counters[value] += count;
+                        }
+                    }
+                    sbg.barrier();
+                }
+                //  Parallel scan on counters to generate offsets in place
+                IntType offset = 0;
+                for (std::uint32_t j = local_id; j < radix_range; j += local_size) {
+                    IntType value = counters[j];
+                    IntType boundary =
+                        sycl::ONEAPI::exclusive_scan(sbg, value, cl::sycl::ONEAPI::plus<IntType>());
+                    counters[j] = offset + boundary;
+                    IntType partial_offset =
+                        sycl::ONEAPI::reduce(sbg, value, cl::sycl::ONEAPI::plus<IntType>());
+                    offset += partial_offset;
+                }
+
+                sbg.barrier();
+                for (std::uint32_t j = local_id; j < group_aligned_size + local_size;
+                     j += local_size) {
+                    bool exists = j < group_aligned_size || local_id < rem;
+                    std::uint8_t c = exists ? cinput[j * radix_count + i] : 0;
+                    IntType local_offset = 0;
+                    std::uint32_t entry = 0;
+                    bool entry_found = false;
+
+                    for (std::uint32_t k = 0; k < local_size; k++) {
+                        bool correct = j < group_aligned_size || k < rem;
+                        std::uint32_t done = sycl::ONEAPI::broadcast(sbg, correct ? 0 : 1, k);
+                        if (done)
+                            break;
+                        std::uint32_t skip = sycl::ONEAPI::broadcast(sbg, entry_found ? 1 : 0, k);
+                        if (skip)
+                            continue;
```

Comment:
Please move it into the loop as conditions.

Reply:
Will try as part of optimization task